src/EventListener/TagListener.php

```diff
@@ -88,14 +105,15 @@ public function onKernelResponse(FilterResponseEvent $event)
             }
         }
 
-        if ($request->isMethodCacheable()) {
+        if ($this->cacheableRule->matches($request, $response)) {
+            // For safe requests (GET and HEAD), set cache tags on response
             $this->symfonyResponseTagger->addTags($tags);
             if (HttpKernelInterface::MASTER_REQUEST === $event->getRequestType()) {
-                // For safe requests (GET and HEAD), set cache tags on response
                 $this->symfonyResponseTagger->tagSymfonyResponse($response);
             }
-        } elseif (count($tags)) {
-            // For non-safe methods, invalidate the tags
+        } elseif (count($tags)
```

Comment:
i would prfer to return early if !count($tags). with your early return if its neither cacheable nor mustinvalidate, this here would be a simple `else` i think.